server/server.go

```diff
@@ -0,0 +1,472 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	epplib "github.com/dotse/epp-lib"
+)
+
+type errorLog interface {
+	Debugf(string, ...interface{})
+	Errorf(string, ...interface{})
+	Infof(string, ...interface{})
+}
+
+// Server is an EPP server.
+type Server struct {
+	// Handler handles commands for a connection. Write the response on rw and
+	// read the command from cmd.
+	Handler func(ctx context.Context, rw *epplib.ResponseWriter, cmd io.Reader)
+
+	// Greeting is called once when a new connection is established and should
+	// write a greeting on rw.
+	Greeting func(ctx context.Context, rw *epplib.ResponseWriter)
+
+	// ConnContext can be used to add metadata to a connection. All future
+	// calls to the Handler will include this context.
+	ConnContext func(ctx context.Context, conn *tls.Conn) (context.Context, error)
+
+	// CloseConnHook can be used to run a function after a connection has been closed,
+	// taking the closed connection as an argument.
+	CloseConnHook func(ctx context.Context, conn *tls.Conn)
+
+	TLSConfig tls.Config
+
+	// Timeout is the total time a connection can stay open on the server.
+	// After this duration the connection is automatically closed.
+	Timeout time.Duration
+
+	// IdleTimeout is how long the connection will stay open without any
+	// activity.
+	IdleTimeout time.Duration
+
+	// WriteTimeout is how long to wait for writes on the response writer.
+	WriteTimeout time.Duration
+
+	// ReadTimeout is how long to wait for reading a command.
+	ReadTimeout time.Duration
+
+	// MaxMessageSize if set will return an error if the incoming request
+	// is bigger than the set size in bytes. 0 indicates no limit.
+	MaxMessageSize int64
+
+	// ErrorLog logs errors when accepting connections, unexpected behavior
+	// from handlers and underlying connection errors.
+	ErrorLog errorLog
+
+	// We keep track of our active connections here. This is guarded by mu.
+	activeConn map[*eppConn]struct{}
+
+	// mu guards activeConn.
+	mu sync.Mutex
+
+	// counts active connections, in created on new connections and decreased
+	// when connections are closed.
+	wg sync.WaitGroup
+
+	// where we accept new connections.
+	listener   epplib.Listener
+	listenerMu sync.RWMutex
+}
+
+// Serve will start a server on the provided listener.
+func (s *Server) Serve(listener epplib.Listener) error {
+	if s.Handler == nil || s.Greeting == nil {
+		panic("Handler and Greeting is required")
+	}
+
+	s.listenerMu.Lock()
+	s.listener = listener
+	s.listenerMu.Unlock()
+
+	s.activeConn = make(map[*eppConn]struct{})
+
+	defer func() {
+		s.mu.Lock()
+
+		for c := range s.activeConn {
+			_ = c.stopAwaitMessage()
+		}
+
+		s.mu.Unlock()
+
+		s.wg.Wait()
+	}()
+
+	for {
+		s.listenerMu.RLock()
```

Comment:
Should we add a comment regarding why this lock is needed?

Reply:
Sure, do you have a suggestion for the comment?

Reply:
I'll have to check.

Reply:
Something like this maybe? (// mutex for listener to not accept new connections when closing the server)

Reply:
Sounds good!